pkg/solana/contract.go

```diff
@@ -92,32 +195,35 @@ func (c *ContractTracker) fetchState(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
+
+	// acquire lock and write to state
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
```

Comment:
Might want to minimize to only what is necessary while holding this lock, and do `PublicKeyFromBytes` and logging before/after.